arch/arm/src/arm/arm_vectors.S

```diff
@@ -136,12 +136,23 @@ arm_vectorirq:
 	mov	sp, r4				/* Restore the possibly unaligned stack pointer */
 #endif
 
+	/* Switch back IRQ mode and return with shadow SPSR */
+
+	mov	r4, #(PSR_MODE_IRQ | PSR_I_BIT)
+	msr	cpsr_c, r4		/* Switch back IRQ mode */
+
 	/* Restore the CPSR, SYS mode registers and return */
 
 	ldr	r0, [sp, #(4*REG_CPSR)]	/* Fetch the return SPSR */
 	msr	spsr_cxsf, r0		/* Set the return mode SPSR */
 
-	ldmia	sp, {r0-r15}^		/* Return */
+	/* Life is simple when everything is IRQ mode */
+
+	mov	r14, sp			/* (IRQ) r14=Register storage area */
+	ldmia	r14!, {r0-r12}		/* Restore common r0-r12 */
+	ldmia	r14, {r13, r14}^	/* Restore user mode r13/r14 */
+	add	r14, r14, #(4*2)	/* (IRQ) r14=address of r15 storage */
+	ldmia	r14, {r15}^		/* Return */
```

Comment:
Back to my original question: why we can't `ldmia r14, {r13, r14, r15}^`instead of 3 lines?

Reply:
As we discussed before, if r13/r14/r15 with suffix ^ are restored, LDM will not restore r13 and r14 to (USR/SYS) mode registers,

```
ldmia   r14, {r13, r14, r15}^

(gdb) i r
sp             0x9f                0x9f
lr             0x1082619c          276980124
pc             0x10800230          0x10800230 <arm_vectorirq+112>
cpsr           0x80000192          -2147483246

(gdb) x/4x 0x1082619c
0x1082619c:     0x108262b0      0x10802c63      0x10802c62      0x2000007f
                       SP           LR             PC             CPSR

**   Restore of r13/r14 will fail with participation of PC (r15):    **

(gdb) i r
sp             0x10826168          0x10826168      <-wrong
lr             0x10800cd3          276827347      <-wrong
pc             0x10802c62          0x10802c62 <uart_write+188>
cpsr           0x2000007f          536871039
```


So we must to restore r13/r14 and r15 separately:


```
ldmia   r14, {r13, r14}^    /* Restore user mode r13/r14 */
add   r14, r14, #(4*2)    /* (IRQ) r14=address of r15 storage */
ldmia   r14, {r15}^     /* Return */

(gdb) i r
sp             0x9f                0x9f
lr             0x108261a4          276980132
pc             0x10800238          0x10800238 <arm_vectorirq+120>
cpsr           0x80000192          -2147483246


(gdb) x/4x 0x108261a4-8
0x1082619c:     0x108262b0      0x10802c63      0x10802c62      0x2000007f

sp             0x108262b0          0x108262b0
lr             0x10802c63          276835427
pc             0x10802c62          0x10802c62 <uart_write+188>
cpsr           0x2000007f          536871039
```





```
^
is an optional suffix, available in ARM state only. You must not use it in User mode or System mode. It has the following purposes:

If the instruction is LDM (with any addressing mode) and reglist contains the PC (R15), in addition to the normal multiple register transfer, the SPSR is copied into the CPSR. This is for returning from exception handlers. Use this only from exception modes.
Otherwise, data is transferred into or out of the User mode registers instead of the current mode registers.
```


https://developer.arm.com/documentation/dui0489/c/Cihcadda

Reply:
Got it. Thank you again for detailed explanation.

Reply:
I will try to do some experiments to see if
```
add r0, sp, #(4 * REG_R13)
ldmia r0, {r13, r14}^
ldmia sp, {r0-r12,r15}^
```
will work. Just for my curiosity.

Reply:
> ldmia sp, {r0-r12,r15}^

reglist of xcp context is different with restore order.